src/common/liboptparse/optparse.c

```diff
@@ -1365,7 +1366,18 @@ int optparse_parse_args (optparse_t *p, int argc, char *argv[])
                                &li, &d, p->posixly_correct)) >= 0) {
         struct option_info *opt;
         struct optparse_option *o;
-        if (c == '?') {
+        if (c == ':') {
+            opt = find_option_by_val (p, d.optopt);
+            if (opt && opt->p_opt->name)
+                (*p->log_fn) ("%s: '--%s' missing argument\n",
+                              fullname,  opt->p_opt->name);
+            else
+                (*p->log_fn) ("%s: '-%c' missing argument\n",
+                              fullname,  d.optopt);
```

Comment:
Isn't the current index into `argv` available in `d.optind`? Can you try instead:
```
(*p->log_fn) ("%s: '%s' missing argument\n", argv[d.optind-1])
```
I think this will end up using the option the user actually provided.